module1/lessons/ruby_object_model_poly.md

```diff
@@ -0,0 +1,220 @@
+
+Also notice that we call `.superclass` on the `Dog` class, not the dog instance. What happens if we call `.superclass` on a dog instance? Try it to find out.
+
+Because our class inherits from `Object`, which inherits from `BasicObject`, we know that any class we create will have those two ancestors. This is where `:new` comes from. Look in the [Ruby Docs BasicObject Page](https://ruby-doc.org/core-2.4.1/BasicObject.html) and you'll see the `:new` defined there.
+
+<section class="dropdown">
+
+### Here is how you might implement a similar structure in JavaScript
+
+```javascript
+class Animal {
+}
+
+class Dog extends Animal {
+}
+
+const dog = new Dog();
+dog instanceof Animal; // true
+Object.getPrototypeOf(Dog); // Animal
+```
+
+</section>
+
+### Mixins and Modules
+
+We've mapped out the inheritance chain for our dog, but what about the modules? What happens if we call `included_modules` on our `Dog` class.
+
+```ruby
+pry(main)> Dog.included_modules
+#=> [AnimalBehavior, PP::ObjectMixin, Kernel]
+```
+
+We can see our `AnimalBehavior` module (along with some others), but we included that in `Animal`, not in `Dog`, so why is it showing up here? `included_modules` shows all modules that were included in *any* superclass, so it won't tell us **where** that module was included. In this case, we get more information if we start at the top of the inheritance chain to figure out where modules first appear (you may get slightly different results depending on what Ruby version you are running):
+
+```ruby
+pry(main)> BasicObject.included_modules
+# => []
+
+pry(main)> Object.included_modules
+# => [PP::ObjectMixin, Kernel]
+
+pry(main)> Animal.included_modules
+# => [AnimalBehavior, PP::ObjectMixin, Kernel]
+
+pry(main)> Dog.included_modules
+# => [AnimalBehavior, PP::ObjectMixin, Kernel]
+```
+
+From this information, we can deduce that `BasicObject` doesn't include any modules, `Object` includes `PP:ObjectMixin` and `Kernel` (you don't need to worry about what those are), and `Animal` includes `AnimalBehavior`.
+
+Our updated diagram:
+
+![Imgur](https://i.imgur.com/f4pszOG.png)
+
+<section class="note">
+
+### Note
+
+JavaScript does not have native support for mixins within its class system.
+
+</section>
+
+## Chart Exercise
+
+Break into small groups.
+
+Using `.class`, `.ancestors`, `.included_modules`, and `.superclass`, diagram the Object Model of these several commonly-used Ruby classes: Hash, Array, String, Integer, and Float.
+
+# The Lookup Chain and Polymorphism
+
+In programming, polymorphism lets us use the same method or property in different classes or objects. Each class or object can have its own special version of a method through a process called **method overriding**.
```

Comment:
I had to reread **method overriding** vs **method overloading** a couple times thinking they were saying the same thing 
- consider changing to **method over*riding*** just to help out any dyslexic students

Reply:
At this point in their education, we would not have talked at all about method overloading so I don't think a student would have this same issue.

Reply:
Instead of changing the formatting for the word method overriding, I decided to add more context.

--

In programming, polymorphism lets us use the same method or property in different classes or objects. Each class or object can have its own special version of a method through a process called **method overriding**, where a local definition of a method overrides the same method from the parent class.